src/algebra/square.lean

```diff
@@ -0,0 +1,201 @@
+lemma square_mul_of_square_of_square {R : Type*} [comm_monoid R] {m n : R} :
+  square m → square n → square (m * n) :=
+begin
+  rintros ⟨c, hc⟩,
+  rintros ⟨d, hd⟩,
+  use (c * d),
+  have : c * d * (c * d) = (c * c) * (d * d),
+  { rw mul_assoc,
+    conv in (d * (c * d)) { rw [mul_comm, mul_assoc], },
+    rw ←mul_assoc, },
+  rw [this, hc, hd],
+end
+
+section factorization
+
+namespace nat
+
+instance : decidable_pred (square : ℕ → Prop)
+| n := decidable_of_iff' _ (nat.exists_mul_self n)
+
+lemma square_iff_factorization_even {m : ℕ} :
+square m ↔ ∀ (p : ℕ), even (m.factorization p) :=
+begin
+  by_cases hm_zero : m = 0,
+  { simp only [hm_zero, square_zero, factorization_zero, finsupp.coe_zero,
+      pi.zero_apply, forall_const, true_iff],
+    use 0,
+    exact mul_zero 2, },
+  { split,
+    { rintros ⟨c, hc⟩,
+      by_cases hc' : c = 0,
+      { rw [hc', mul_zero] at hc,
+        rw ←hc,
+        simp only [factorization_zero, finsupp.coe_zero, pi.zero_apply, forall_const],
+        use 0,
+        exact mul_zero 2, },
+      { intros p,
+        rw [←hc, nat.factorization_mul hc' hc', finsupp.coe_add, ←two_mul],
+        simp [even], }, },
+    { intros hp,
+      use m.factorization.prod (λ p i, p ^ (i / 2)),
+      rw ← finsupp.prod_mul,
+      conv { to_rhs, rw ← nat.factorization_prod_pow_eq_self hm_zero,},
+      apply finsupp.prod_congr,
+      intros x hx,
+      rcases hp x with ⟨c, hc⟩,
+      rw [←pow_add, ←two_mul, hc],
+      congr,
+      simp, }, },
+end
+
+/-- The property of being a square is multiplicative. The ← direction can be generalized
+to an arbitrary commutative monoid. See `square_mul_of_square_of_square`. Similarly,
+this lemma could be generalized to an arbitrary unique factorization domain, but we make use
+of `nat.factorization` in this proof. -/
+lemma square_mul {m n : ℕ} (hmn : m.coprime n) :
+  square (m * n) ↔ square m ∧ square n :=
+begin
+  split,
+  { rw [square_iff_factorization_even, square_iff_factorization_even],
+    rw square_iff_factorization_even,
+    intros hmn',
+    split,
+    { intros p,
+      by_cases hp : p ∈ m.factors,
+      { rw ← factorization_eq_of_coprime_left hmn hp, exact hmn' p, },
+      { rw [←nat.factors_count_eq, list.count_eq_zero_of_not_mem hp],
+        use zero,
+        exact mul_zero 2, }, },
+    { intros p,
+      by_cases hp : p ∈ n.factors,
+      { rw ← factorization_eq_of_coprime_right hmn hp, exact hmn' p, },
+      { rw [←nat.factors_count_eq, list.count_eq_zero_of_not_mem hp],
+        use zero,
+        exact mul_zero 2, }, }, },
+  { rintros ⟨hm, hn⟩, exact square_mul_of_square_of_square hm hn, },
+end
+
+lemma square_prime_pow_iff_pow_even : ∀ (p i : ℕ), nat.prime p → (square (p ^ i) ↔ even i) :=
+begin
+  intros p i hp,
+  split,
+  { rintros ⟨s, hs⟩,
+    have : s ∣ p ^ i,
+      calc s ∣ s * s : dvd_mul_left s s
+        ... = p ^ i : hs,
+    rw nat.dvd_prime_pow hp at this,
```

Comment:
```suggestion
    have := dvd_mul_left s s,
    rw [hs, nat.dvd_prime_pow hp] at this,
```

Reply:
Thanks! Still not good a these transitive things....